<a href="https://colab.research.google.com/github/nissrinayy/deeplearning/blob/main/week%201/Regression_Infrared_Dataset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Tugas Week 1 Deep Learning 🚀🚀

In [ ]:
import pandas as pd

from google.colab import files
uploaded = files.upload()  # Upload manual jika belum diunggah

df = pd.read_csv("Infrared.csv")  # Load dataset
pd.set_option('display.max_columns', None)
df.head()  # Lihat isi data


Saving Infrared.csv to Infrared (2).csv


Gender    Age                  Ethnicity  T_atm  Humidity  Distance  \
0    Male  41-50                      White   24.0      28.0       0.8   
1  Female  31-40  Black or African-American   24.0      26.0       0.8   
2  Female  21-30                      White   24.0      26.0       0.8   
3  Female  21-30  Black or African-American   24.0      27.0       0.8   
4    Male  18-20                      White   24.0      27.0       0.8   

   T_offset1  Max1R13_1  Max1L13_1  aveAllR13_1  aveAllL13_1    T_RC1  \
0     0.7025    35.0300    35.3775      34.4000      34.9175  34.9850   
1     0.7800    34.5500    34.5200      33.9300      34.2250  34.7100   
2     0.8625    35.6525    35.5175      34.2775      34.8000  35.6850   
3     0.9300    35.2225    35.6125      34.3850      35.2475  35.2075   
4     0.8950    35.5450    35.6650      34.9100      35.3675  35.6025   

   T_RC_Dry1  T_RC_Wet1  T_RC_Max1    T_LC1  T_LC_Dry1  T_LC_Wet1  T_LC_Max1  \
0    34.9850    34.7625    35.0325  35.3375    35.3375    34.4850    35.3775   
1    34.6325    34.6400    34.7425  34.5600    34.5375    34.3500    34.5750   
2    35.6675    35.6150    35.7175  35.5025    35.5025    35.2950    35.5300   
3    35.2000    35.1175    35.2250  35.5950    35.5950    35.3275    35.6125   
4    35.4750    35.5700    35.6400  35.6400    35.6400    35.0775    35.6675   

      RCC1     LCC1  canthiMax1  canthi4Max1  T_FHCC1  T_FHRC1  T_FHLC1  \
0  34.7850  34.4650     35.3775      35.3375  33.5775  33.4775  33.3725   
1  34.3225  34.2400     34.7400      34.7150  34.0325  34.0550  33.6775   
2  35.3575  35.0925     35.7175      35.6825  34.9000  34.8275  34.6475   
3  34.9100  35.1700     35.6125      35.5950  34.4400  34.4225  34.6550   
4  35.3550  35.1200     35.6650      35.6475  35.0900  35.1600  34.3975   

   T_FHBC1  T_FHTC1  T_FH_Max1  T_FHC_Max1   T_Max1    T_OR1  T_OR_Max1  \
0  33.4925  33.0025    34.5300     34.0075  35.6925  35.6350    35.6525   
1  33.9700  34.0025    34.6825     34.6600  35.1750  35.0925    35.1075   
2  34.8200  34.6700    35.3450     35.2225  35.9125  35.8600    35.8850   
3  34.3025  34.9175    35.6025     35.3150  35.7200  34.9650    34.9825   
4  34.6700  33.8275    35.4175     35.3725  35.8950  35.5875    35.6175   

   aveOralM  
0     36.59  
1     37.19  
2     37.34  
3     37.09  
4     37.04

In [ ]:
print(df.info())  # Menampilkan tipe data


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1020 entries, 0 to 1019
Data columns (total 34 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Gender       1020 non-null   object 
 1   Age          1020 non-null   object 
 2   Ethnicity    1020 non-null   object 
 3   T_atm        1020 non-null   float64
 4   Humidity     1020 non-null   float64
 5   Distance     1018 non-null   float64
 6   T_offset1    1020 non-null   float64
 7   Max1R13_1    1020 non-null   float64
 8   Max1L13_1    1020 non-null   float64
 9   aveAllR13_1  1020 non-null   float64
 10  aveAllL13_1  1020 non-null   float64
 11  T_RC1        1020 non-null   float64
 12  T_RC_Dry1    1020 non-null   float64
 13  T_RC_Wet1    1020 non-null   float64
 14  T_RC_Max1    1020 non-null   float64
 15  T_LC1        1020 non-null   float64
 16  T_LC_Dry1    1020 non-null   float64
 17  T_LC_Wet1    1020 non-null   float64
 18  T_LC_Max1    1020 non-null   float64
 19  RCC1  

# PreProcessing Data

In [ ]:
!pip install --upgrade scikit-learn

In [ ]:
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.model_selection import train_test_split
import numpy as np

**Menghilangkan Missing Vallues** pada fitur distance (memiliki 1018 dari 1020)



In [ ]:
# Hapus data dengan nilai NaN di kolom Distance
df_clean = df.dropna(subset=['Distance'])

**Melakukan Encoding**
terdapat 3 feature yang bertipe kategori (Gender, Age, Ethnicity), sisanya numerik. Dilakukan encoding untuk fitur katerogikal

In [ ]:
# One-hot encoding untuk kategori
categorical_features = ['Gender', 'Age', 'Ethnicity']
encoder = OneHotEncoder(sparse_output=False, drop='first')
encoded_categorical = encoder.fit_transform(df_clean[categorical_features])


**Melakukan normalisasi** untuk fitur bertipe numerik

In [ ]:
numerical_features = [col for col in df_clean.columns if col not in categorical_features + ['aveOralM']]
scaler = StandardScaler()
scaled_numerical = scaler.fit_transform(df_clean[numerical_features])

In [ ]:
# Gabungkan semua fitur
X = np.hstack((scaled_numerical, encoded_categorical))
y = df_clean['aveOralM'].values


split data

In [ ]:
# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


## Model PyTorch

In [ ]:
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader, TensorDataset
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

Konversi data ke PyTorch

In [ ]:
# Konversi data ke Tensor untuk PyTorch
X_train_tensor = torch.tensor(X_train, dtype=torch.float32)
y_train_tensor = torch.tensor(y_train, dtype=torch.float32).view(-1, 1)
X_test_tensor = torch.tensor(X_test, dtype=torch.float32)
y_test_tensor = torch.tensor(y_test, dtype=torch.float32).view(-1, 1)

# Dataset PyTorch
dataset_train = TensorDataset(X_train_tensor, y_train_tensor)
dataloader_train = DataLoader(dataset_train, batch_size=32, shuffle=True)


Membuat model Pytorch

In [ ]:
# Model PyTorch
class RegressionModel(nn.Module):
    def __init__(self, input_dim):
        super(RegressionModel, self).__init__()
        self.model = nn.Sequential(
            nn.Linear(input_dim, 64),
            nn.ReLU(),
            nn.Linear(64, 32),
            nn.ReLU(),
            nn.Linear(32, 1)
        )

    def forward(self, x):
        return self.model(x)

# Inisialisasi model PyTorch
input_dim = X_train.shape[1]
model_pytorch = RegressionModel(input_dim)
criterion = nn.MSELoss()
optimizer = optim.Adam(model_pytorch.parameters(), lr=0.001)

# Membuat Model TensorFlow

In [ ]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

In [ ]:
# Model TensorFlow
model_tf = keras.Sequential([
    keras.Input(shape=(input_dim,)),  # Tentukan input di luar Dense
    layers.Dense(64, activation='relu'),
    layers.Dense(32, activation='relu'),
    layers.Dense(1)
])


model_tf.compile(optimizer='adam', loss='mse', metrics=['mae'])

# Evaluasi Model

1. **Mean Squared Error (MSE)**
    MSE menghitung rata-rata dari kuadrat selisih antara nilai sebenarnya (yi) dan nilai prediksi (y^)
2. **Root Mean Squared Error (RMSE)**
    RMSE adalah akar dari MSE, sehingga memiliki satuan yang sama dengan data aslinya.
3.  **R-squared (r2)**
    mengukur seberapa baik model dapat menjelaskan variabilitas dalam data

    r2=1 -> Model sempurna (semua prediksi benar).

    r2=0 -> Model tidak lebih baik daripada rata-rata.
    
    r2<0 -> Model lebih buruk dari hanya mengambil rata-rata data.

In [ ]:
# Evaluasi Model PyTorch
model_pytorch.eval()
with torch.no_grad():
    y_pred_pytorch = model_pytorch(X_test_tensor).numpy()

mse_pytorch = mean_squared_error(y_test, y_pred_pytorch)
rmse_pytorch = np.sqrt(mse_pytorch)
r2_pytorch = r2_score(y_test, y_pred_pytorch)


print(f"PyTorch MSE: {mse_pytorch:.4f}")
print(f"PyTorch RMSE: {rmse_pytorch:.4f}")
print(f"PyTorch R^2: {r2_pytorch:.4f}")

PyTorch MSE: 1379.3236
PyTorch RMSE: 37.1392
PyTorch R^2: -6933.0071


In [ ]:
# Evaluasi Model TensorFlow
y_pred_tf = model_tf.predict(X_test)

mse_tf = mean_squared_error(y_test, y_pred_tf)
rmse_tf = np.sqrt(mse_tf)
r2_tf = r2_score(y_test, y_pred_tf)

print(f"TensorFlow MSE: {mse_tf:.4f}")
print(f"TensorFlow RMSE: {rmse_tf:.4f}")
print(f"TensorFlow R^2: {r2_tf:.4f}")

7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step 
TensorFlow MSE: 1386.4308
TensorFlow RMSE: 37.2348
TensorFlow R^2: -6968.7359
